notebooks/Binary Classification with the UCI Credit Card Payment Default Dataset.ipynb

```diff
@@ -0,0 +1,599 @@
+    "        \"Demographic parity difference\": (lambda x: group_selection_rate(y_true, x, group).range, True),\n",
+    "        \"Demographic parity ratio\": (lambda x: group_selection_rate(y_true, x, group).range_ratio, True),\n",
+    "        \"Error rate difference\": (\n",
+    "            lambda x: metric_by_group(balanced_accuracy_score, y_true, x, group).range, True),\n",
+    "        \"Equal opportunity difference\": (lambda x: group_recall_score(y_true, x, group).range, True),\n",
+    "        \"Equalized odds difference\": (lambda x: group_equalized_odds_diff(y_true, x, group), True),\n",
+    "        \"Group AUC difference\": (lambda x: group_roc_auc_score(y_true, x, group).range, False),\n",
+    "        \"Overall AUC\": (lambda x: roc_auc_score(y_true, x), False)\n",
+    "        }\n",
+    "    df_dict = {}\n",
+    "    for metric_name, (metric_func, use_preds) in metrics_dict.items():\n",
+    "        df_dict[metric_name] = [metric_func(preds) if use_preds else metric_func(scores) \n",
+    "                                for model_name, (preds, scores) in models_dict.items()]\n",
+    "    return pd.DataFrame.from_dict(df_dict, orient=\"index\", columns=models_dict.keys())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We calculate several disparity metrics below:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Metrics\n",
+    "models_dict = {\"Unmitigated\": (test_preds, test_scores)}\n",
+    "get_metrics_df(models_dict, Y_test, A_str_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Mitigating Demographic Disparities with Postprocessing Techniques"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We attempt to mitigate the disparities in the `lightgbm` prediction using the `fairlearn` postprocessing technique `ThersholdOptimizer`. This technique will find a suitable threshold for the scores (class probabilities) produced by the `lightgbm` model such that the demographic disparities are minimized. For this exercise, we focus on the `equalized odds` disparity metric."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "postprocess_est = ThresholdOptimizer(\n",
+    "    estimator=model,\n",
+    "    constraints=\"equalized_odds\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "balanced_idx1 = df_train[Y_train==1].index\n",
+    "pp_train_idx = balanced_idx1.union(Y_train[Y_train==0].sample(n=balanced_idx1.size, random_state=1234).index)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_train_balanced = df_train.loc[pp_train_idx, :]\n",
```

Comment:
These cells could use a comment, so I don't have to think too hard about what's going on.